libs/spock-shaded/src/main/java/org/codice/spock/extension/builtin/DeFinalizeClassLoader.java

```diff
@@ -0,0 +1,168 @@
+ * loaded by the parent classloader.
+ */
+public class DeFinalizeClassLoader extends ClassLoader {
+  private final ClassPool pool;
+  private final Set<String> filters;
+
+  /**
+   * Constructs a new classloader for the specified Spock test specification class. The class is
+   * used to retrieve the {@link DeFinalize} annotations in order to identify which classes or
+   * packages should be definalized as they are being loaded.
+   *
+   * <p>The classloader used to load this classloader class will be defined as its parent
+   * classloader.
+   *
+   * @param specClass the Spock test specification class
+   */
+  public DeFinalizeClassLoader(Class<?> specClass) {
+    super(DeFinalizeClassLoader.class.getClassLoader());
+    this.pool = new ClassPool(false);
+    this.pool.appendClassPath(new LoaderClassPath(getParent()));
+    this.pool.appendSystemPath();
+    this.filters =
+        Stream.concat(
+                Stream.of(specClass.getAnnotationsByType(DeFinalize.class))
+                    .map(DeFinalize::value)
+                    .flatMap(Stream::of)
+                    .map(Class::getName),
+                Stream.of(specClass.getAnnotationsByType(DeFinalize.class))
+                    .map(DeFinalize::packages)
+                    .flatMap(Stream::of))
+            .collect(Collectors.toSet());
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * <p>Loads the class with the specified <a href="#name">binary name</a>.
+   *
+   * <p>The class will always be loaded first from its parent classloader. It will be returned as is
+   * if the class is from one of the reserved packages (see above). Otherwise, the class will be
+   * reloaded and optionally definalize if the class match a requested class or package from one of
+   * the {@link DeFinalize} annotations of the Spock test specification class. The protection domain
+   * from the version of the class loaded by the parent classloader will be re-used when defining
+   * the reloaded class.
+   *
+   * @param name the binary name of the class to load
+   * @param resolve <code>true</code> to resolve the class; <code>false</code> otherwise
+   * @return the resulting class object
+   * @throws ClassNotFoundException if the class could not be found
+   */
+  @Override
+  protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+    synchronized (getClassLoadingLock(name)) {
+      Class<?> clazz = findLoadedClass(name);
+
+      if (clazz == null) {
+        clazz = super.loadClass(name, resolve); // always load it from our parent first
+        if (!name.startsWith("java.")
+            && !name.startsWith("javax.")
+            && !name.startsWith("sun.")
+            && !name.startsWith("org.xml.")
+            && !name.startsWith("org.junit.")) {
+          try {
+            boolean definalize = filters.contains(name);
+
+            if (!definalize) {
+              for (int i = name.indexOf('.'); i > 0; i = name.indexOf('.', ++i)) {
```

Comment:
✏️ Probably worth adding a quick comment above this for-loop. It took me a second to realize this was the package-matching code.

Reply:
😄 Will do. I am always for adding comments.